### PR TITLE
ARROW-6631: [C++] Do not build any compression libraries by default in C++ build

### PR DESCRIPTION
**Pull request by wesm**
From the look of the diff, I would say that configuring a comprehensive build looks too complicated to me (it's only relevant for developers and packagers, though, so more a maintainability nuisance). Perhaps ARROW-6585 can help with this

Also fixes a unit test that only fails when all compressors are disabled. I needed to add a `Codec::IsAvailable` function to determine whether a codec's support has been built, which seems useful anyway.

### CODE REVIEW

**Comment by nealrichardson** [comment]
Re: complexity, what if the cmake var was like `ARROW_COMPRESSION=bz2,zlib,zstd,lz4,snappy,brotli` and could take some subset of those, or `all` for all of them?

**Comment by fsaintjacques** [comment]
Single options toggling is better since it'll integrate with IDE and cmake-aware tools.

**Comment by wesm** [comment]
It'd be nice to have a shortcut. I suggest we tackle adding such shortcuts holistically. I think it's still helpful to have on/off flags in some cases

**Comment by wesm** [comment]
Looks like the wheels are failing for an unrelated reasons (something having to do with protobuf in Homebrew). I'll fix the Windows issue

**Comment by nealrichardson** [comment]
Appveyor failure looks legit too: https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/27961055/job/4rdh31hvlbh6r184

**Comment by kou** [comment]
@ursabot crossbow submit -g linux

**Review by kou — APPROVED** [review]
+1 for Linux packages.

File: docs/source/developers/cpp.rst [hunk 1]
```diff
@@ -174,6 +169,12 @@ boolean flags to ``cmake``.
 * ``-DARROW_PARQUET=ON``: Apache Parquet libraries and Arrow integration
 * ``-DARROW_PLASMA=ON``: Plasma Shared Memory Object Store
 * ``-DARROW_PLASMA_JAVA_CLIENT=ON``: Build Java client for Plasma
+* ``-DARROW_WITH_BZ2=ON``: Build support for BZ2 compression
```
Comment (pitrou) [inline docs/source/developers/cpp.rst]: Can we keep the list more or less alphabetical?

File: docs/source/developers/python.rst [hunk 2]
```diff
@@ -287,6 +292,8 @@ to ``OFF``:
 * ``ARROW_ORC``: Support for Apache ORC file format
 * ``ARROW_PARQUET``: Support for Apache Parquet file format
 * ``ARROW_PLASMA``: Shared memory object store
+* ``ARROW_WITH_{ZLIB, ZSTD, LZ4, SNAPPY, BROTLI}``: Build support for
```
Comment (pitrou) [inline docs/source/developers/python.rst]: And BZ2?

File: docs/source/developers/python.rst [hunk 3]
```diff
@@ -270,6 +270,11 @@ Now build and install the Arrow C++ libraries:
          -DARROW_FLIGHT=ON \
          -DARROW_GANDIVA=ON \
          -DARROW_ORC=ON \
+         -DARROW_WITH_ZLIB=ON \
```
Comment (pitrou) [inline docs/source/developers/python.rst]: BZ2 as well?

**Review by pitrou — APPROVED** [review]
LGTM, just a nit

File: docs/source/developers/cpp.rst (hunk 1 above)
Comment (wesm) [reply docs/source/developers/cpp.rst]: done

File: docs/source/developers/python.rst (hunk 2 above)
Comment (wesm) [reply docs/source/developers/python.rst]: Not needed for Parquet but I'm adding here anyhow

File: docs/source/developers/python.rst (hunk 3 above)
Comment (wesm) [reply docs/source/developers/python.rst]: Adding -- I had only listed the compressors used for Parquet

**Pull request closed without merging**